### PR TITLE
ref: Next iteration of span API

### PR DESCRIPTION
**Pull request by untitaker**

### CODE REVIEW

File: sentry_sdk/tracing.py [hunk 1]
```diff
@@ -123,7 +132,31 @@ def __repr__(self):
...
+        # type: () -> Span
+        hub = self.hub or sentry_sdk.Hub.current
+
+        _, scope = hub._stack[-1]
+        old_span = scope.span
+        scope.span = self
+        self._context_manager_state = (hub, scope, old_span)
+        return self
+
+    def __exit__(self, ty, value, tb):
```
Comment (kamilogorek) [inline sentry_sdk/tracing.py]: I assume `value` is some kind of error here? Or exception parameter? Thats why it fails the transaction?

File: sentry_sdk/tracing.py [hunk 2]
```diff
@@ -183,29 +221,77 @@ def to_traceparent(self):
...
 
-    def finish(self):
-        self.timestamp = datetime.now()
-        if self._finished_spans is not None:
-            self._finished_spans.append(self)
+    def finish(self, hub=None):
+        # type: (Optional[sentry_sdk.Hub]) -> Optional[str]
+        hub = hub or self.hub or sentry_sdk.Hub.current
+
+        if self.timestamp is None:
```
Comment (kamilogorek) [inline sentry_sdk/tracing.py]: I think it should short-circuit here, so same transaction cannot be sent multiple times.

```
if self.timestamp is not None:
  return
```

File: sentry_sdk/tracing.py [hunk 3]
```diff
@@ -303,7 +391,8 @@ def record_http_request(hub, url, method):
                     span.set_data(k, v)
 
 
-def maybe_create_breadcrumbs_from_span(hub, span):
+def _maybe_create_breadcrumbs_from_span(hub, span):
+    # type: (sentry_sdk.Hub, Span) -> None
     if span.op == "redis":
         hub.add_breadcrumb(type="redis", category="redis", data=span._tags)
     elif span.op == "http" and span.is_success():
```
Comment (kamilogorek) [inline sentry_sdk/tracing.py]: You don't want failed http requests to be breadcrumbed as well?

File: sentry_sdk/tracing.py (hunk 1 above)
Comment (untitaker) [reply sentry_sdk/tracing.py]: `value` is an exception object while `ty` is its type and `tb` a traceback object.

File: sentry_sdk/tracing.py (hunk 2 above)
Comment (untitaker) [reply sentry_sdk/tracing.py]: I agree, let's do that

File: sentry_sdk/tracing.py (hunk 3 above)
Comment (untitaker) [reply sentry_sdk/tracing.py]: The old Python SDK didn't do that, and I don't want to change this UI for now.

**Pull request merged**